tests/instrumentation-framework/src/main/java/org/hawkular/apm/tests/dockerized/TestScenarioRunner.java

```diff
@@ -0,0 +1,230 @@
+     * @param apmServerPort - port at which Hawkular APM server will be started
+     * @throws DockerCertificateException
+     */
+    public TestScenarioRunner(String apmVersion, int apmServerPort) throws DockerCertificateException {
+        this.apmVersion = apmVersion;
+        this.apmServerPort = apmServerPort;
+
+        this.objectMapper = new ObjectMapper();
+        apmServer = new ApmMockServer();
+        apmServer.setPort(apmServerPort);
+        apmServer.setShutdownTimer(60*60*1000);
+    }
+
+    /**
+     * Run a test scenario
+     *
+     * @param testScenario
+     * @return number of successful test cases
+     */
+    public int run(TestScenario testScenario) {
+        if (testScenario.getEnvironment().getDockerCompose() != null &&
+                testScenario.getEnvironment().getImage() != null) {
+            throw new IllegalArgumentException("Ambiguous environment: defined docker image and docker-compose");
+        }
+
+        System.out.println("Starting test scenario: " + testScenario);
+        int successfulTestCases = 0;
+
+        for (TestCase test: testScenario.getTests()) {
+            if (test.isSkip()) {
+                continue;
+            }
+
+            TestEnvironmentExecutor testEnvironmentExecutor = testEnvironmentExecutor(testScenario);
+
+            try {
+                runTestCase(testScenario, test, testEnvironmentExecutor);
+                successfulTestCases++;
+            } catch (TestFailException ex) {
+                System.out.println("Test case failed: " + ex.toString());
+                System.out.println(ex.getMessage());
+                ex.printStackTrace();
+            }
+
+            testEnvironmentExecutor.close();
+        }
+
+        return successfulTestCases;
+    }
+
+    /**
+     * Run single test case
+     *
+     * @param testScenario Test scenario of the test case
+     * @param testCase Test case to run
+     * @param testEnvironmentExecutor Initialized environment
+     * @throws TestFailException
+     */
+    private void runTestCase(TestScenario testScenario, TestCase testCase,
+                             TestEnvironmentExecutor testEnvironmentExecutor) throws TestFailException {
+
+        System.out.println("Executing test case: " + testCase);
+
+        String environmentId = null;
+        try {
+            if (testEnvironmentExecutor instanceof DockerComposeExecutor) {
+                ((DockerComposeExecutor) testEnvironmentExecutor).createNetwork();
+            }
+
+            /**
+             * Start APM server
+             */
+            apmServer.setHost("0.0.0.0");
+            apmServer.setTraces(new ArrayList<>());
+            apmServer.run();
+
+            /**
+             * Run a container and wait
+             */
+            environmentId = testEnvironmentExecutor.run(testScenario.getEnvironment());
+            Thread.sleep(testScenario.getEnvironment().getInitWaitSeconds()*1000);
```

Comment:
This will sleep no matter what, right? I see two problems with this approach: first, if the environment inits very fast, you'll always end up wasting `(initWaitSeconds*1000)-actualInitTime`. And second, on some environments, you might need more than just a few seconds to start. Ideally, you'd do the `run`, then `join`, failing if it hasn't joined after `N` seconds (timeout).